Do not allow creating replicated table with inconsistent merge params

canhld94 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Do not allow creating replicated table with inconsistent merge params

See https://github.com/ClickHouse/ClickHouse/pull/56724#issuecomment-1812541732


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

canhld94 committed:
save all merge params to zookeeper

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 commented:
Not implement for graphite merge tree yet. My idea for graphite merge tree is to save as 128 sip hash value of graphite config in zookeeper. Any better idea is appreciated.

canhld94 committed:
calculate hash for graphite merge params

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

robot-ch-test-poll commented:
<!-- automatic status comment for PR #56833 from canhld94/ClickHouse:ch_compare_merge_params -->
*This is an automated comment for commit e08fffbff71c9ea97bcc89a0edba6ce679063f83 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56833/e08fffbff71c9ea97bcc89a0edba6ce679063f83/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6888266575">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56833/e08fffbff71c9ea97bcc89a0edba6ce679063f83/ci_running.html">❌ failure</a></td></tr>
<tbody>
</table>

canhld94 committed:
add graphite params hash to zookeeper + fix tests

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
install new graphite for testing

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
fix backward incompatibility

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
minor fix test

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 commented:
Kindly remind @tavplubix [THANKS]

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeTreeTableMetadata.cpp:
```diff
@@ -36,6 +39,21 @@ static String formattedASTNormalized(const ASTPtr & ast)
 
+static bool peekString(const char * s, ReadBufferFromString & buf)
+{
+    auto * current_pos = buf.position();
+    for (; *s; ++s)
+    {
+        if (buf.eof() || *buf.position() != *s)
+        {
+            buf.position() = current_pos;
```
It's unsafe to use a previously saved position after checking for eof, because `eof()` might call `nextImpl()` which might refill the buffer with a new bunch of data or even reallocate the buffer. See `PeekableReadBuffer`

canhld94 replied:
> because eof() might call nextImpl() which might refill the buffer with a new bunch of data or even reallocate the buffer
Yes, I'm aware of this so I only implement `peekString` for `ReadBufferFromString`. But now look at `ReadBufferFromString` code seem it doesn't work as I thought, still have internal buffer. Then this implementation is incorrect.

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeTreeTableMetadata.cpp:
```diff
@@ -121,6 +150,19 @@ void ReplicatedMergeTreeTableMetadata::write(WriteBuffer & out) const
+        if (!version_column.empty())
+            out << "version column: " << version_column << "\n";
+        if (!is_deleted_column.empty())
+            out << "is delete column: " << is_deleted_column << "\n";
```
```suggestion
            out << "is_deleted column: " << is_deleted_column << "\n";
```

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeTreeTableMetadata.cpp:
```diff
@@ -142,34 +184,54 @@ void ReplicatedMergeTreeTableMetadata::read(ReadBuffer & in)
     if (in.eof())
         data_format_version = 0;
-    else if (checkString("data format version: ", in))
+    else if (peekString("data format version: ", in))
```
It's better to keep `checkString` if possible because peeking is too complex

canhld94 replied:
Then some must be extra careful when adding something new to metadata, e.g. two consecutive field name cannot have same prefix. I guess it's ok, don't expect to touch this code too often.

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeTreeTableMetadata.cpp:
```diff
@@ -121,6 +150,19 @@ void ReplicatedMergeTreeTableMetadata::write(WriteBuffer & out) const
     if (!constraints.empty())
         out << "constraints: " << constraints << "\n";
+
+    if (merge_params_version >= REPLICATED_MERGE_TREE_METADATA_WITH_ALL_MERGE_PARAMETERS)
+    {
+        out << "merge parameters format version: " << merge_params_version << "\n";
```
Let's add a compatibility setting that disables the new version. Otherwise, it will be impossible to downgrade after you create a new table or add a new replica. See https://github.com/ClickHouse/ClickHouse/pull/42869#issuecomment-1311785870 
We will enable it after a few releases

canhld94 replied:
I add `merge_params_version` here to enforce backward compatibility. Integration test has some test for backward compatibility, and I also tested backward compatibility locally.

I tested in two scenarios: (1) 1 replica has new version and one replica has old version, and (2) create table in new version, then downgrade the cluster. Both are working as expected.

The key point is that deserialising replicated metadata doesn't check for EOF, so I think as long as we only add new fields to the end then it's always backward compatible.

wdyt?

canhld94 committed:
Update src/Storages/MergeTree/ReplicatedMergeTreeTableMetadata.cpp

Co-authored-by: Alexander Tokmakov <tavplubix@gmail.com>

canhld94 committed:
remove peekString and add more comments

- peekString doesn't always work even for ReadBufferFromString
- more comment re. backward compatibility

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

tavplubix commented:
AST fuzzer (debug) - https://github.com/ClickHouse/ClickHouse/issues/57077
Integration tests (asan) [3/4] - https://github.com/ClickHouse/ClickHouse/issues/57854
Integration tests (tsan) [6/6] - https://github.com/ClickHouse/ClickHouse/pull/57863
Stress test (tsan) - looks like CI infrastructure issue
